### ISSUE TITLE
Problem with one FE0F test

### ISSUE THREAD

**Issue by Carbon225** [adraffy/ens-normalize.js#22]
Hi, I'm the author of the [NameHash ens-normalize-python](https://github.com/namehash/ens-normalize-python) library. I tried updating our test JSONs to the latest ones but found one failing case.

Here is the test case I have trouble with:
```json
{
    "name": "0x✌️🏼",
    "tokenized": [
        {
            "type": "valid",
            "cps": [
                48,
                120
            ]
        },
        {
            "type": "emoji",
            "emoji": [
                9996,
                65039
            ],
            "input": [
                9996,
                65039
            ],
            "cps": [
                9996
            ]
        },
        {
            "type": "disallowed",
            "cp": 127996
        }
    ],
    "error": true,
    "comment": "Malicious FE0F"
}
```

With some modifications to our source code I reproduced the expected outcome (this made other tests fail though):
```python
[TokenValid(cps=[48, 120], type='valid'),
 TokenEmoji(emoji=[9996, 65039], input=[9996, 65039], cps=[9996], type='emoji'),
 TokenDisallowed(cp=127996, type='disallowed')]
```

This is the tokenization I get with our original implementation:
```python
[TokenValid(cps=[48, 120], type='valid'),
 TokenEmoji(emoji=[9996, 127996], input=[9996, 65039, 127996], cps=[9996, 127996], type='emoji')]
```

What I want to figure out is whether it's an issue with our code. I believe the entire [9996, 65039, 127996] emoji sequence should be matched. [9996, 127996] is an allowed emoji in the spec. Since matching FE0Fs is optional, the whole sequence is matched in our code. Your implementation seems to break matching after the FE0F, and the skin tone is reported as disallowed.

From the test case comment I assume it's some kind of "attack".
Can you confirm which is the expected behavior?

**Comment by adraffy** [comment]
Yeah, I was planning on reaching out about this one but I forgot.  This test slipped through the cracks in my code due to how I was compressing the emoji.  (I had a quirk flag which would account for this case but not during parsing.) 

My emoji compression code was collapsing the following 2 emoji: `A FE0F` and `A B`, into the following structure: `A` node, which allowed `FE0F`, is a valid emoji (eg. `A` or `A FE0F`), and also branched to `B` node, which didn't allow` FE0F` and is a valid emoji (eg. `A B`).  But this incorrectly allows `A FE0F B` to parse as valid.  (Note: my previous implementation did disallow `A FE0F FE0F B`.)

While `FE0F` are optional, it should only be optional in the negative sense — input `C D` matches emoji `C FE0F D` but input `C FE0F D` does not match emoji `C D`.  The regex example illustrates how the optionality of the FE0F implies that it can only be missing (not extra).  

I will add an extra sentence to ENSIP-15 to make this more clear.  I'll will also add a few more incorrectly-inserted FE0F test cases in the Unicode 15.1 update.

My intention is that we should only accept fully-qualified emoji (how they appear in spec.json) or those emoji with `FE0F` removed (eg. normalized or partially-normalized).  Everything else should be rejected (like randomly inserting `FE0F` between a flag emoji).

This change essentially makes parsing slightly more strict and matches my original intention.

**Comment by adraffy** [comment]
Here is the algorithm that I like for this is as it is extremely simple.

* https://github.com/adraffy/ens-normalize.js/blob/main/src/lib.js#L124
* https://github.com/adraffy/ENSNormalize.java/blob/main/lib/src/main/java/io/github/adraffy/ens/ENSIP15.java#L69
* https://github.com/adraffy/ENSNormalize.cs/blob/main/ENSNormalize/ENSIP15.cs#L160

**Comment by adraffy** [comment]
Another improvement I found while porting C# and Java was in `determine_group()` where the [error was generated](https://github.com/adraffy/ens-normalize.js/blob/main/src/lib.js#L406C42-L406C42).   In the case where the next codepoint doesn't exist in any group, the produced error can be fine-tuned by checking if that character exists in any group.
* If no groups, then the character was a NFC composition that was input as valid decomposed parts but itself isn't canonical form isn't valid.
* If any group, then it's mixed script.

**Comment by Carbon225** [comment]
Ok great, I see now. Back when I was implementing this I basically had to reverse your code and that's how I interpreted this optional matching. Thanks for clarifying.

**Comment by adraffy** [comment]
I apologize about that.  The C# implementation is probably a better reference now as it's fully-typed.

**Comment by Carbon225** [comment]
> input `C D` matches emoji `C FE0F D` but input `C FE0F D` does not match emoji `C D`

I really like this example. All is fixed on our end.

I also noticed the behavior for empty names was changed. They used to be disallowed. Currently, `ens_normalize(empty_string)` should return `empty_string` without any errors, right?

**Comment by Carbon225** [comment]
I'll close this since it's resolved. If you could please take a look at https://github.com/ensdomains/docs/pull/152#pullrequestreview-1550092257
